src/anndata/_core/sparse_dataset.py

```diff
@@ -250,30 +250,47 @@ def slice_as_int(s: slice, l: int) -> int:
 def get_compressed_vectors(
     x: BackedSparseMatrix, row_idxs: Iterable[int]
 ) -> tuple[Sequence, Sequence, Sequence]:
-    slices = [slice(*(x.indptr[i : i + 2])) for i in row_idxs]
-    data = np.concatenate([x.data[s] for s in slices])
-    indices = np.concatenate([x.indices[s] for s in slices])
-    indptr = list(accumulate(chain((0,), (s.stop - s.start for s in slices))))
+    indptr_slices = [slice(*(x.indptr[i : i + 2])) for i in row_idxs]
+    # HDF5 cannot handle out-of-order integer indexing
+    if isinstance(x.data, ZarrArray):
+        as_np_indptr = np.concatenate(
+            [np.arange(s.start, s.stop) for s in indptr_slices]
+        )
+        data = x.data[as_np_indptr]
+        indices = x.indices[as_np_indptr]
+    else:
+        data = np.concatenate([x.data[s] for s in indptr_slices])
+        indices = np.concatenate([x.indices[s] for s in indptr_slices])
+    indptr = list(accumulate(chain((0,), (s.stop - s.start for s in indptr_slices))))
     return data, indices, indptr
 
 
 def get_compressed_vectors_for_slices(
     x: BackedSparseMatrix, slices: Iterable[slice]
 ) -> tuple[Sequence, Sequence, Sequence]:
-    indptr_sels = [x.indptr[slice(s.start, s.stop + 1)] for s in slices]
-    data = np.concatenate([x.data[s[0] : s[-1]] for s in indptr_sels])
-    indices = np.concatenate([x.indices[s[0] : s[-1]] for s in indptr_sels])
+    indptr_indices = [x.indptr[slice(s.start, s.stop + 1)] for s in slices]
+    indptr_limits = [(i[0], i[-1]) for i in indptr_indices]
+    # HDF5 cannot handle out-of-order integer indexing
+    if isinstance(x.data, ZarrArray):
+        indptr_int = np.concatenate(
+            [np.arange(start, stop) for start, stop in indptr_limits]
+        )
+        data = x.data[indptr_int]
+        indices = x.indices[indptr_int]
+    else:
+        data = np.concatenate([x.data[start:stop] for start, stop in indptr_limits])
+        indices = np.concatenate(
+            [x.indices[start:stop] for start, stop in indptr_limits]
+        )
     # Need to track the size of the gaps in the slices to each indptr subselection
-    total = indptr_sels[0][0]
-    offsets = [total]
-    for i, sel in enumerate(indptr_sels[1:]):
-        total = (sel[0] - indptr_sels[i][-1]) + total
-        offsets.append(total)
-    start_indptr = indptr_sels[0] - offsets[0]
+    offsets = [indptr_indices[0][0]]
+    for (_, stop0), (start1, _) in pairwise(indptr_limits):
+        offsets.append((start1 - stop0) + offsets[-1])
```

Comment:
Should be something like this, right?

```suggestion
    gaps = (start1 - stop0 for (_, stop0), (start1, _) in pairwise(indptr_limits))
    offsets = accumulate(chain([indptr_limits[0][0]] gaps))
```

As said above, you could also do `indptr_slices` instead of `indptr_limits`, then we’d get

```diff
-indptr_int = np.concatenate([np.arange(start, stop) for start, stop in indptr_limits])
+indptr_int = np.concatenate([np.arange(s.start, s.stop) for s in indptr_slices])
```

and

```diff
-offsets = accumulate(chain([indptr_limits[0][0]] gaps))
+offsets = accumulate(chain([indptr_slices[0].start] gaps))
```

which I think is nicer

Reply:
The accumulation iterator is great.  It cleans up some code below as well.